File: playwright_py/pages/CheckOutPage.py
from typing import Final
from playwright.sync_api import Page, expect
from .base_page import BasePage
from ..constants import urls
from ..utils import DataGenerator 


class CheckOutPage(BasePage):
    """Page object for the Checkout Step One page."""
    page_url_fragment: Final[str] = urls.checkout_step_one

    def __init__(self, page: Page):
        super().__init__(page)
        self.page = page
        self.first_name_input = page.locator('[data-test="firstName"]')
        self.last_name_input = page.locator('[data-test="lastName"]')
        self.postal_code_input = page.locator('[data-test="postalCode"]')
        self.continue_btn = page.locator('[data-test="continue"]')
        self.checkout_form = page.locator('.checkout_info') 
        self.cancel_btn = page.locator('[data-test="cancel"]')
        self.page_title = page.locator('[data-test="title"]')
        self.data_gen = DataGenerator()

    def check_is_on_checkout_page(self):
        """Verify that we are on the Checkout Step One page."""
        expect(self.page).to_have_url(lambda url: self.page_url_fragment in url)
        expect(self.first_name_input).to_be_visible()

    async def fill_checkout_form(self, user_data: dict = None):
        """Fill the checkout form with provided user data."""
        if not user_data:
            user_data = self.data_gen.get_random_user(user_role="standard_user")
        await self.first_name_input.fill(user_data['firstName'])
        await self.last_name_input.fill(user_data['lastName'])
        await self.postal_code_input.fill(user_data['postalCode'])
        await self.continue_btn.click()

File: playwright_ts/pages/CheckOutPage.ts
import { _, url, User, UserRole } from 'sdk_automation';
import { BasePage, CartPage } from '../pages';
import { Page } from '@playwright/test';

export class CheckOutPage extends BasePage {
  static readonly selectors = {
    firstNameInput: '[data-test="firstName"]',
    backToProductsBtn: '[data-test="back-to-products"]',
    lastNameInput: '[data-test="lastName"]',
    postalCodeInput: '[data-test="postalCode"]',
    continueButton: '[data-test="continue"]',
    cancelButton: '[data-test="cancel"]',
    pageTitle: '[data-test="title"]',
    checkoutForm: '.checkout_info',
    finishButton: '[data-test="finish"]',
  };

  readonly page: Page;
  readonly pageUrlStepOne = url.checkoutStepOne;
  readonly pageUrlStepTwo = url.checkoutStepTwo;
  readonly pageUrlComplete = url.checkoutComplete;
  readonly backToProductsBtn;
  readonly firstNameInput;
  readonly lastNameInput;
  readonly postalCodeInput;
  readonly continueButton;
  readonly cancelButton;
  readonly pageTitle;
  readonly checkoutForm;
  readonly finishButton;
  readonly cartPage: CartPage;

  constructor(page: Page) {
    super(page);
    this.page = page;
    this.cartPage = new CartPage(page);
    this.backToProductsBtn = page.locator(CheckOutPage.selectors.backToProductsBtn);
    this.firstNameInput = page.locator(CheckOutPage.selectors.firstNameInput);
    this.lastNameInput = page.locator(CheckOutPage.selectors.lastNameInput);
    this.postalCodeInput = page.locator(CheckOutPage.selectors.postalCodeInput);
    this.continueButton = page.locator(CheckOutPage.selectors.continueButton);
    this.cancelButton = page.locator(CheckOutPage.selectors.cancelButton);
    this.checkoutForm = page.locator(CheckOutPage.selectors.checkoutForm);
    this.pageTitle = page.locator(CheckOutPage.selectors.pageTitle);
    this.finishButton = page.locator(CheckOutPage.selectors.finishButton);
  }

  async navigateToCheckoutForm(): Promise<void> {
    try {
      await this.checkUrl(this.cartPage.pageUrl);
      await this.cartPage.checkoutBtn.waitFor({ state: 'visible' });
      await this.cartPage.checkoutBtn.click();
      await this.checkUrl(this.pageUrlStepOne);
    } catch (error) {
      throw new Error(`[navigateToCheckoutForm]: Failed to navigate to checkout form: ${error}`);
    }
  }

  async fillCheckoutForm(userData?: User): Promise<void> {
    try {
      const data = userData ??
        _.getRandomUser?.(UserRole.standard_user) ?? {
          firstName: 'John',
          lastName: 'Doe',
          postalCode: '12345',
        };

      await this.firstNameInput.fill(data.firstName);
      await this.lastNameInput.fill(data.lastName);
      await this.postalCodeInput.fill(data.postalCode);
      await this.continueButton.click();
      await this.checkUrl(this.pageUrlStepTwo);
    } catch (error) {
      throw new Error(`[fillCheckoutForm]: ${error}`);
    }
  }

  async finishCheckout(): Promise<void> {
    await this.finishButton.click();
    await this.page.waitForURL(url.checkoutComplete);
  }
}
